torch_xla/experimental/xla_sharding.py

```diff
@@ -319,45 +319,35 @@ def _get_sharding_type(partition_spec: Tuple[Union[int, None]],
   return sharding_type
 
 
-def _get_tile_assignment(mesh: Mesh,
-                         partition_spec: Tuple[Union[int, None]]) -> List[int]:
-  # Use Torch.tensor here to make use of the torch.transpose_
-  mesh_list_tensor = torch.tensor(mesh.get_logical_mesh().tolist())
-  partition_spec_list = list(partition_spec)
-  for i in range(len(partition_spec_list)):
-    if partition_spec_list[i] == None:
-      partition_spec_list[i] = i
-  # We currently do not support partition_spec like [0, None, 1, 3]. The None at partition_spec[1]
-  # suggested that we want to replicate on Mesh[1], hence we can't use Mesh[1] in
-  # partition_spec[2]
-  assert torch.unique(
-      torch.tensor(partition_spec_list)).size()[0] == len(partition_spec_list)
-  return mesh_list_tensor.permute(partition_spec_list).tolist()
-
-
-def _get_group_assignment(sharding_type: ShardingType,
-                          partition_spec: Tuple[Union[int, None]],
-                          tile_assignment: List) -> Tuple[List, List]:
+def _get_tile_assignment(mesh: Mesh, partition_spec: Tuple[Union[int, None]]) -> List[int]:
+  mesh_size = 1 if type(mesh.mesh_shape) == int else len(mesh.mesh_shape)
+  spec_size = 1 if type(partition_spec) == int else len(partition_spec)
+  if (None not in partition_spec) and (mesh_size == spec_size):
+    return mesh.get_logical_mesh().transpose(partition_spec).tolist()
+  # Tile permutation is not necessary for partial replication.
+  return mesh.get_logical_mesh().tolist()
+
+
+def _get_group_assignment(
+    sharding_type: ShardingType, mesh: Mesh,
+    partition_spec: Tuple[Union[int, None]]) -> Tuple[List, List]:
   group_assignment = list()
   replication_groups = list()
-  mesh_shape_list = list(torch.tensor(tile_assignment).size())
   if sharding_type is ShardingType.PARTIAL:
     # Shard across groups and replicate within subgroups; replicated dims
     # will be used to group replication devices.
     tile_dims = [d for d in partition_spec if d is not None]
-    replicated_dims = set(range(len(mesh_shape_list))) - set(tile_dims)
+    replicated_dims = set(range(len(mesh.mesh_shape))) - set(tile_dims)
 
-    group_list = [np.array(tile_assignment)]
+    group_list = [np.array(mesh.get_logical_mesh().tolist())]
     for d in tile_dims:
       _group_list = list()
       for group_members in group_list:
-        _group_list += np.split(group_members, mesh_shape_list[d], d)
+        _group_list += np.split(group_members, mesh.mesh_shape[d], d)
       group_list = _group_list
     replication_groups = [group.flatten().tolist() for group in group_list]
 
-    group_tile_shape = mesh_shape_list
-    for d in replicated_dims:
-      group_tile_shape[d] = 1
+    group_tile_shape = [mesh.mesh_shape[d] if d is not None else 1 for d in partition_spec]
     group_assignment = np.arange(len(replication_groups)).reshape(
         tuple(group_tile_shape)).tolist()
   return group_assignment, replication_groups
```

Comment:
can you add a comment here or in the function declaration regarding how to interpret `group_assignment` and `replication group`?

Reply:
Done